src/Middleware/HttpLogging/src/W3CLogger.cs

```diff
@@ -0,0 +1,130 @@
+                        elements[BitOperations.Log2((int)W3CLoggingFields.Host)] = val.Trim();
+                        break;
+                    case "User-Agent":
+                        // User-Agent can have whitespace - we replace whitespace characters with the '+' character
+                        elements[BitOperations.Log2((int)W3CLoggingFields.UserAgent)] = Regex.Replace(val.Trim(), @"\s", "+");
+                        break;
+                    case nameof(HeaderNames.Referer):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.Referer)] = val.Trim();
+                        break;
+                    case nameof(DateTime):
+                        DateTime dt = DateTime.Parse(val, CultureInfo.InvariantCulture);
+                        elements[BitOperations.Log2((int)W3CLoggingFields.Date)] = dt.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);
+                        elements[BitOperations.Log2((int)W3CLoggingFields.Time)] = dt.ToString("HH:mm:ss", CultureInfo.InvariantCulture);
+                        // We estimate time elapsed by diffing the current time with the time at which the middleware processed the request/response.
+                        // This will represent the time in whole & fractional milliseconds.
+                        var elapsed = DateTime.Now.Subtract(dt);
+                        elements[BitOperations.Log2((int)W3CLoggingFields.TimeTaken)] = elapsed.TotalMilliseconds.ToString(CultureInfo.InvariantCulture);
+                        break;
+                    case nameof(HeaderNames.Server):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.ServerName)] = val.Trim();
+                        break;
+                    case nameof(ConnectionInfo.RemoteIpAddress):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.ClientIpAddress)] = val.Trim();
+                        break;
+                    case nameof(ConnectionInfo.LocalIpAddress):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.ServerIpAddress)] = val.Trim();
+                        break;
+                    case nameof(ConnectionInfo.LocalPort):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.ServerPort)] = val.Trim();
+                        break;
+                    case nameof(HttpContext.User):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.UserName)] = val.Trim();
+                        break;
+                    case nameof(HeaderNames.Cookie):
+                        // Cookie can have whitespace - we replace whitespace characters with the '+' character
+                        elements[BitOperations.Log2((int)W3CLoggingFields.Cookie)] = Regex.Replace(val.Trim(), @"\s", "+");
+                        break;
+                    default:
+                        break;
+                }
+            }
+            StringBuilder sb = new StringBuilder();
```

Comment:
`ValueStringBuilder`?

Reply:
What's the advantage of `ValueStringBuilder`?

Reply:
`ValueStringBuilder` is struct, so less allocation.